server/src/main/java/com/genymobile/scrcpy/Controller.java

```diff
@@ -145,18 +147,28 @@ private boolean injectKeycode(int action, int keycode, int repeat, int metaState
     }
 
     private boolean injectChar(char c) {
-        String decomposed = KeyComposition.decompose(c);
-        char[] chars = decomposed != null ? decomposed.toCharArray() : new char[]{c};
-        KeyEvent[] events = charMap.getEvents(chars);
-        if (events == null) {
-            return false;
-        }
-        for (KeyEvent event : events) {
-            if (!device.injectEvent(event)) {
+        if (options.useADBKeyboard()) {
+            // Process latin keys the same way in order to provide same reaction speed.
+            try {
+                Process process = Runtime.getRuntime().exec("am broadcast -a ADB_INPUT_CHARS --eia chars " + String.valueOf((int) c));
+                return process.waitFor() == 0;
```

Comment:
You should send the broadcast programmatically (to avoid execute a new process and add latency for every letter).

I guess this involves retrieving an `IActivityManager` by calling [`ActivityManager.getService()`](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/ActivityManager.java#4070) (by introspection) (or [`ActivityManagerNative.getService()`](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/ActivityManagerNative.java#50) on older versions), and calling [`broadcastIntent()`](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/IActivityManager.aidl#120).

Reply:
Is that method stable enough? I only have A9 devices and there is something made about activities in A10. I'm not Android developer though.